packages/beacon-node/src/chain/historicalState/historicalState.ts

```diff
@@ -0,0 +1,160 @@
+import {HistoricalStateRegenMetrics, IBinaryDiffCodec, StateArchiveStrategy} from "./types.js";
+import {replayBlocks} from "./utils/blockReplay.js";
+import {DiffLayers} from "./diffLayers.js";
+import {BinaryDiffVCDiffCodec} from "./utils/binaryDiffVCDiffCodec.js";
+import {getDiffState} from "./utils/diff.js";
+
+const codec: IBinaryDiffCodec = new BinaryDiffVCDiffCodec();
+
+export async function getHistoricalState(
+  {slot}: {slot: Slot},
+  {
+    db,
+    logger,
+    config,
+    metrics,
+    diffLayers,
+    pubkey2index,
+  }: {
+    config: BeaconConfig;
+    db: IBeaconDb;
+    pubkey2index: PubkeyIndexMap;
+    logger: Logger;
+    diffLayers: DiffLayers;
+    metrics?: HistoricalStateRegenMetrics;
+  }
+): Promise<Uint8Array | null> {
+  const regenTimer = metrics?.regenTime.startTimer();
+  const epoch = computeEpochAtSlot(slot);
+  const strategy = diffLayers.getArchiveStrategy(slot);
+  logger.debug("Fetching state archive", {strategy, slot, epoch});
+
+  switch (strategy) {
+    case StateArchiveStrategy.Snapshot: {
+      const loadStateTimer = metrics?.loadStateTime.startTimer();
+      const state = await db.stateSnapshotArchive.getBinary(slot);
+      loadStateTimer?.();
+      regenTimer?.({strategy: StateArchiveStrategy.Snapshot});
+      return state;
+    }
+    case StateArchiveStrategy.Diff: {
+      const {diffState} = await getDiffState({slot, skipSlotDiff: false}, {db, metrics, logger, diffLayers, codec});
+      regenTimer?.({strategy: StateArchiveStrategy.Diff});
+
+      return diffState;
+    }
+    case StateArchiveStrategy.BlockReplay: {
+      const {diffState, diffSlots} = await getDiffState(
+        {slot, skipSlotDiff: false},
+        {db, metrics, logger, diffLayers, codec}
+      );
+      if (!diffState) return null;
+
+      const state = replayBlocks(
+        {toSlot: slot, lastFullSlot: diffSlots[diffSlots.length - 1], lastFullState: diffState},
+        {config, db, metrics, pubkey2index}
+      );
+
+      regenTimer?.({strategy: StateArchiveStrategy.BlockReplay});
+
+      return state;
+    }
+  }
+}
+
+export async function putHistoricalSate(
+  {slot, state}: {slot: Slot; state: Uint8Array},
+  {
+    db,
+    logger,
+    metrics,
+    diffLayers,
+  }: {
+    db: IBeaconDb;
+    logger: Logger;
+    metrics?: HistoricalStateRegenMetrics;
+    diffLayers: DiffLayers;
+  }
+): Promise<void> {
+  const epoch = computeEpochAtSlot(slot);
+  const strategy = diffLayers.getArchiveStrategy(slot);
+
+  switch (strategy) {
+    case StateArchiveStrategy.Snapshot: {
+      metrics?.stateSnapshotSize.set(state.byteLength);
+      await db.stateSnapshotArchive.putBinary(slot, state);
+      logger.verbose("State stored as snapshot", {
+        epoch,
+        slot,
+      });
+      break;
+    }
+    case StateArchiveStrategy.Diff: {
+      const {diffState} = await getDiffState({slot, skipSlotDiff: true}, {db, metrics, logger, diffLayers, codec});
+
+      if (!diffState) return;
```

Comment:
need some logs here

Reply:
Yes a lot of comments need to be added, working on it.